lib/matrix.js

```diff
@@ -52,11 +58,351 @@ var init = function(exports){
+                    }
+                }
+                return stateEvents;
+            }
+            else {
+                return [this.rooms[roomId].state[type][stateKey]];
+            }
+        },
+        
+        /*
+         * Return a single state MatrixEvent from the store for the given roomId
+         * and type.  stateKey is optional; if missing, is assumed to be blank.
+         */
+        getStateEvent: function(roomId, type, stateKey) {
+            if (stateKey === undefined) {
+                for (var i in this.rooms[roomId].state[type]) {
+                    if (this.rooms[roomId].state[type].hasOwnProperty(i)) {
+                        return this.rooms[roomId].state[type][i];
+                    }
+                }
+            }
+            else {
+                return this.rooms[roomId].state[type][stateKey];
+            }
+        },
+        
+        /*
+         * Adds a list of arbitrary MatrixEvents into the store.
+         * If the event is a state event, it is also updates state.
+         */
+        setEvents: function(events) {
+            for (var i = 0; i < events.length; i++) {
+                var event = events[i].event;
+                if (event.type === "m.presence") {
+                    this.setPresenceEvents([events[i]]);
+                    continue;
+                }
+                var roomId = event.room_id;
+                if (this.rooms[roomId] === undefined) {
+                    this.rooms[roomId] = {};
+                }                
+                if (this.rooms[roomId].timeline === undefined) {
+                    this.rooms[roomId].timeline = [];
+                }
+                if (event.state_key !== undefined) {
+                    this.setStateEvents([events[i]]);
+                }
+                this.rooms[roomId].timeline.push(events[i]);
+            }
+        },
+        
+        /*
+         * Get the timeline of events for a given room
+         * TODO: ordering!
+         */
+        getEvents: function(roomId) {
+            return this.room[roomId].timeline;
+        },
+        
+        setPresenceEvents: function(presenceEvents) {
+            for (var i = 0; i < presenceEvents.length; i++) {
+                var matrixEvent = presenceEvents[i];
+                this.presence[matrixEvent.event.user_id] = matrixEvent;
+            }
+        },
+        
+        getPresenceEvents: function(userId) {
+            return this.presence[userId];
+        },
+        
+        getRoomList: function() {
+            var roomIds = [];
+            for (var roomId in this.rooms) {
+                if (this.rooms.hasOwnProperty(roomId)) {
+                    roomIds.push(roomId);
+                }
+            }
+            return roomIds;
+        },
+        
+        // TODO
+        setMaxHistoryPerRoom: function(maxHistory) {},
+        
+        // TODO
+        reapOldMessages: function() {},
```

Comment:
May be worth commenting these 2 functions out for now: we risk people assuming they are implemented otherwise, and spend hours thinking "why isn't it reaping, my heap is still huuuuge" until they check the source and think "Well *thanks*."